securesystemslib/signer/_key.py

```diff
@@ -180,22 +215,80 @@ def from_dict(cls, keyid: str, key_dict: Dict[str, Any]) -> "SSlibKey":
     def to_dict(self) -> Dict[str, Any]:
         return self._to_dict()
 
+    def _from_pem(self) -> "PublicKeyTypes":
+        """Helper to load public key instance from PEM-formatted keyval."""
+        public_bytes = self.keyval["public"].encode("utf-8")
+        return load_pem_public_key(public_bytes)
+
+    @staticmethod
+    def _hash_algo(name) -> Type["HashAlgorithm"]:
+        """Helper to return hash algorithm class for name."""
+        algos = {
+            "sha224": SHA224,
+            "sha256": SHA256,
+            "sha384": SHA384,
+            "sha512": SHA512,
+        }
+        return algos[name]
+
     def verify_signature(self, signature: Signature, data: bytes) -> None:
         try:
-            if not sslib_keys.verify_signature(
-                self.to_securesystemslib_key(),
-                signature.to_dict(),
-                data,
+            sig = bytes.fromhex(signature.signature)
+
+            if CRYPTO_IMPORT_ERROR:
+                if self.scheme == "ed25519":
+                    # Verify using vendored ed25519 implementation
+                    pub = bytes.fromhex(self.keyval["public"])
+                    checkvalid(sig, data, pub)
+                    return
+
+                raise exceptions.UnsupportedLibraryError(CRYPTO_IMPORT_ERROR)
+
+            key: PublicKeyTypes
+            if self.scheme in [
+                "rsassa-pss-sha224",
+                "rsassa-pss-sha256",
+                "rsassa-pss-sha384",
+                "rsassa-pss-sha512",
+                "rsa-pkcs1v15-sha224",
+                "rsa-pkcs1v15-sha256",
+                "rsa-pkcs1v15-sha384",
+                "rsa-pkcs1v15-sha512",
+            ]:
+                key = cast(RSAPublicKey, self._from_pem())
+                padding_name, algo_name = self.scheme.split("-")[1:]
+                algo = self._hash_algo(algo_name)()
+                padding: AsymmetricPadding
+                if padding_name == "pss":
+                    padding = PSS(mgf=MGF1(algo), salt_length=PSS.AUTO)
+                else:
+                    padding = PKCS1v15()
+                key.verify(sig, data, padding, algo)
+
+            elif self.scheme in ["ecdsa-sha2-nistp256", "ecdsa-sha2-nistp384"]:
+                key = cast(EllipticCurvePublicKey, self._from_pem())
+                algo_name = f"sha{self.scheme[-3:]}"
+                algo = self._hash_algo(algo_name)()
+                key.verify(sig, data, ECDSA(algo))
+
+            elif self.scheme in ["ed25519"]:
+                public_bytes = bytes.fromhex(self.keyval["public"])
+                key = Ed25519PublicKey.from_public_bytes(public_bytes)
+                key.verify(sig, data)
+
+            else:
+                raise ValueError(f"unknown scheme '{self.scheme}'")
```

Comment:
I tried different variants here: One, where I didn't dissect the scheme string to get hash and padding objects, but bluntly added a separate elif block per scheme; or another one, where I hid the verification argument assembly in a helper and called `key.verify` only once using some `*args` magic.

I opted for this one here, because it felt like a good trade-off between readability, succinctness and mypy-friendlyness. That said, I'm happy to take any code style advice.

Reply:
Turns out, the refactored code in `SSlibSigner.sign` looks 90% the same. This means, having helpers that dissect the scheme string in order to call the right pyca/cryptography functions definitely makes sense and should be available to SSlibKey and SSlibSigner. And IIRC other signers use the same scheme strings and pyca/cryprography too.

I'll try to deduplicate here.

Reply:
> having helpers that dissect the scheme string in order to call the right pyca/cryptography functions definitely makes sense and should be available to SSlibKey and SSlibSigner.

They should actually be available to other signers too, e.g. _gcp_signer:

https://github.com/secure-systems-lab/securesystemslib/blob/135567fa04f10d0c6a4cd32eb45ce736e1f50a93/securesystemslib/signer/_gcp_signer.py#L161-L172
 
Let's move on with some duplicate helpers here, and ticketize this issue for later.